### DISCUSSION TITLE
Visual Studio Code questions

### DISCUSSION THREAD

**Discussion by dsplaisted** [pybricks/support#1626] [General]
I'm experimenting with PyBricks and Visual Studio code for my team for next year.  I've followed @FLL-Team-24277's [setup instructions](https://github.com/FLL-Team-24277/Master-Program-Fall-2023/) and have some basic things working.

I have some questions about how to make things work better:

- Sometimes I get error squiggles from Pylance indicating that something isn't defined, even though the code works when I run it.  This seemed to happen intermittently for `XboxController`, and is currently showing up for `Number`.  Is there a way to fix this?
- I'd like to use types for my lists so I can get better autocomplete.  I can add `from typing import List` and then do something like `self.measurements: List[PathMeasurement] = []` and I get the autocomplete, but when I run the program I get `ImportError: no module named 'typing'`.  Would it be possible to make this work?  As I understand it the typing library is just for autocomplete and other editor things, and doesn't do anything at runtime.
- When there's an error running the program which is output to the Terminal, the line with the error shows up as highlighted in VS code.  Is there a way to clear that error once it's been fixed?

For background, I'm mainly a C# / .NET programmer, so I don't know a whole lot about Python.

**Comment by dlech** [comment]
> I'd like to use types for my lists so I can get better autocomplete. I can add from typing import List and then do something like self.measurements: List[PathMeasurement] = [] and I get the autocomplete, but when I run the program I get ImportError: no module named 'typing'. Would it be possible to make this work? As I understand it the typing library is just for autocomplete and other editor things, and doesn't do anything at runtime.

There are a couple of things you can do here.

For lists and other builtin types, you don't actually need to use the typing module (assuming you selected Python version in VS Code is >= 3.9). You can just write:

```python
self.measurements: list[PathMeasurement] = []
```

It you really need something from the `typing` module, you need to make sure to not import it at runtime. Here is a little hack to do that.

```python
import micropython

if not micropython:
    import typing
```

Then you won't get the import error because the import is skipped at runtime but you can use it for type hints only. 

Note: I used `micropython` so that this is sort of self documenting, but you could really use any expression that evaluates to falsy at runtime that the code completion tools would not know the value of. (I.e. `if False:` probably won't work because the static analysis tools know that the import will never run because `False` is immutable).

**Comment by dlech** [comment]
> * Sometimes I get error squiggles from Pylance indicating that something isn't defined, even though the code works when I run it.  This seemed to happen intermittently for `XboxController`, and is currently showing up for `Number`.  Is there a way to fix this?

If it is intermittent, probably not much we can do. If something is missing completely or there is a reproducible bug, feel free to open an issue for it with exact step=by-step details on how to reproduce the problem.

**Comment by dlech** [comment]
> * When there's an error running the program which is output to the Terminal, the line with the error shows up as highlighted in VS code.  Is there a way to clear that error once it's been fixed?

It looks like there is a long-standing feature request in VS Code for something like this: https://github.com/microsoft/vscode/issues/66982

In the meantime, I think the only thing you can do to make them go away is run your program again.

**Comment by dsplaisted** [comment]
Thanks a bunch for the help.  If I keep seeing error squiggles I'll try to figure out if there's a way to reproduce it.
